rust/trace-dump/src/chrome_trace.rs

```diff
@@ -29,58 +26,175 @@ pub enum OutputFormat {
 #[derive(Debug)]
 pub enum Error {
     Io(IOError),
-    Json(serde_json::Error)
+    Json(serde_json::Error),
+    DecodingFormat(String)
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct ChromeTraceArrayEntryArgs {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    xi_sample_id: Option<usize>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    payload: Option<TracePayloadT>,
+}
+
+/// This is the struct that represents a single entry within an array of
+/// samples.  Vec<ChromeTraceArrayEntry> would represent a single complete
+/// trace.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct ChromeTraceArrayEntry {
+    name: StrCow,
+    cat: StrCow,
+    ph: StrCow,
+    #[serde(rename = "ts")]
+    ts_us: u64,
+    pid: u64,
+    tid: u64,
+    args: Option<ChromeTraceArrayEntryArgs>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(untagged)]
+enum ChromeTraceArrayEntries {
+    Array(Vec<ChromeTraceArrayEntry>)
 }
 
 fn to_us(ns: u64) -> u64 {
     ns / 1000
 }
 
+fn to_ns(us: u64) -> u64 {
+    us * 1000
+}
+
 fn event_type(sample: &Sample, begin: bool) -> &'static str {
     match sample.sample_type {
         SampleType::Instant => "i",
         SampleType::Duration => if begin { "B" } else { "E" },
     }
 }
 
-fn sample_to_json(sample: &Sample, begin: bool) -> serde_json::Value {
-    json!({
-        "cat": sample.categories.join(","),
-        "name": sample.name, 
-        "ph": event_type(sample, begin),
-        "ts": to_us(if begin { sample.start_ns } else { sample.get_end_ns() }),
-        "pid": sample.pid,
-        "tid": sample.tid,
-        "args": {
-            "payload": sample.payload
+fn sample_to_entry(sample: &Sample, begin: bool) -> ChromeTraceArrayEntry {
+    ChromeTraceArrayEntry {
+        name: sample.name.clone(),
+        cat: StrCow::from(sample.categories.join(",")),
+        ph: StrCow::from(event_type(sample, begin)),
+        ts_us: to_us(if begin { sample.start_ns } else { sample.get_end_ns() }),
+        pid: sample.pid,
+        tid: sample.tid,
+        args: Some(ChromeTraceArrayEntryArgs {
+            xi_sample_id: Some(sample.sample_id),
+            payload: sample.payload.clone(),
+        })
+    }
+}
+
+impl<'a, A> From<A> for ChromeTraceArrayEntries where A: IntoIterator<Item = &'a Sample> {
+    fn from(samples: A) -> Self {
+        // Worst case is every sample is a duration.  if we were to use an
+        // original capacity of samples.len() then we'd inevitably encounter
+        // at least 1 duration & thus the underlying Vec would double the
+        // capacity anyway.  Alternatively, we could scan the vector to count
+        // the number of durations & add that to samples.len() for the exact
+        // final capacity.
+        let samples_iter = samples.into_iter();
+        let mut result = Vec::with_capacity(samples_iter.size_hint().0 * 2);
```

Comment:
I'm curious to hear your thinking around vec buffer sizes, since I've noticed you go out of your way a few times to try and pick an initial capacity. My personal inclination would be to use `with_capacity` when I know the initial capacity exactly or I know it's large, and otherwise not sweat a few reallocs; and if I'm worried about size I can always `shrink_to_fit` when I'm done.

Do you think there's a tangible advantage to trying to make an accurate estimate up front? This is more for my edification, not really a review comment. 🙂

Reply:
If I have a good guess or a reasonable upper-bound, then I tend to provide a capacity.  I tend to not do so when I know there's going to be a large variance in the upper bound.  In this specific case I know it's going to end up allocating exactly 2x anyway using the naiive way so I just do it upfront instead of via a realloc.  The reason I know it's going to be 2x anyway is that it's likely going to be at least len + 1 since there's probably going to be a duration sample.  Thus if I allocate a capacity of len I'll end up with 2*len since vector implementations typically 2x when they need to grow (& I just double-checked that rust does indeed).

shrink_to_fit is good when you know you're going to have a lot of garbage (i.e. after clearing a large vector that's sticking around).